rag-foundation/scripts/main.py

```diff
@@ -0,0 +1,217 @@
+
+    def answer(self, query: str, top_k: int = 5) -> tuple[str, list[str]]:
+        # Generate openai code to answer the query
+        result = self.retrieve(query, top_k=top_k)
+        context_list = [node.text for node in result.nodes]
+        context = "\n\n".join(context_list)
+
+        self.prompt_template = (
+            f"""Question: {query}\n\nGiven context: {context}\n\nAnswer:"""
+        )
+
+        if not self.model:
+            raise ValueError("Model not found. Please initialize the model first.")
+        try:
+            response = self.model.invoke(self.prompt_template)
+        except Exception as e:
+            raise Exception(f"Error in calling the model: {e}")
+        return response.content, context_list
+
+
+def main(
+    data_path: Path = Path("data/qasper-test-v0.3.json"),
+    output_path: Path = Path("predictions.jsonl"),
+    mode: str = "sparse",
+    force_index: bool = False,
+    print_context: bool = False,
+    chunk_size: int = 200,
+    top_k: int = 5,
+    retrieval_only: bool = False,
+):
+    # Generate doc string
+    """
+    Args:
+        data_path: Path to the qasper data file.
+        output_path: Path to save the predictions.
+        mode: Mode of the vector store. Choose either `sparse` or `semantic`.
+        force_index: Whether to force indexing the documents.
+        print_context: Whether to print the context.
+        chunk_size: Chunk size for splitting the documents.
+        top_k: Number of top k documents to retrieve.
+        retrieval_only: Whether to retrieve only.
+    Returns:
+        None
+    """
+    # Load the data
+    raw_data = json.load(open(data_path, "r", encoding="utf-8"))
+
+    question_ids, predicted_answers, predicted_evidences = [], [], []
+
+    # NOTE: qasper has many papers, each paper has multiple sections
+    # we will loop through each paper, gather the full text of each section
+    # and prepare the documents for the vector store
+    # and answer the query
+    for _, values in raw_data.items():
+        # for each paper in qasper
+        documents = []
+
+        for section in values["full_text"]:
+            # for each section in the paper
+            documents += section["paragraphs"]
+
+        # initialize the vector store
+        # and rag pipeline
+        # Remember to force_index=True if you want to override the existing index
+        vector_store = prepare_vector_store(
+            documents, mode=mode, force_index=force_index, chunk_size=chunk_size
+        )
+
+        # NOTE: Should design your own template
+        prompt_template = """Question: {}\n\nGiven context: {}\n\nAnswer:"""
+
+        rag_pipeline = RAGPipeline(vector_store, prompt_template=prompt_template)
+
+        for q in values["qas"]:
+            # for each question in the paper
+            query = q["question"]
+            question_ids.append(q["question_id"])
+
+            # NOTE: If you just want to retrieve the top 5 relevant documents
```

Comment:
Modify to: "If you just want to retrieve the top_k relevant documents..."

Reply:
fixed in [63fad35](https://github.com/Cinnamon/ai-bootcamp-2024/pull/2/commits/63fad35519d68e0436f079e2378440a149878e8b)

Reply:
Thank you @cin-kay.